[Security Solution][Entity Analytics][PrivMon] Add label support to JSON API for privileged user monitoring

abhishekbhatia1710 opened the pull request:
## Summary

This PR adds comprehensive label support to the JSON API for privileged user monitoring, enabling users to create, update, and manage labels through the API while maintaining compatibility with existing CSV upload functionality.

## Key Features

- API Label Support: Users can now include labels in create/update requests via the JSON API
- Source Tracking: Each label maintains its source (api, csv, index_sync) for proper management
- Intelligent Merging: API labels are intelligently merged with existing CSV labels without duplicates
- Source-Aware Deletion: Delete operations only remove API labels when users have other sources
- Backward Compatibility: Existing CSV upload functionality remains unchanged

## Testing Steps

### 1. Create User with Labels via API
POST /api/entity_analytics/monitoring/users with labels in entity_analytics_monitoring.labels

### 2. Upload User with Same Label via CSV
Upload CSV with same user and label

Expected Result: User should have both labels with different sources

### 3. Delete User via JSON API
DELETE /api/entity_analytics/monitoring/users/{user-id}

Expected Result: Only API label should be removed, CSV label should remain

### 4. Delete User via CSV Upload
Remove user from CSV and re-upload

Expected Result: Only CSV label should be removed, API label should remain

## Acceptance Criteria

✅ User uploaded via CSV with label and created via API with same label has both records with different sources
✅ Deleting via JSON API only removes API source labels, preserving CSV labels  
✅ Deleting via CSV upload preserves API labels

Fixes: [EA-13185](https://github.com/elastic/security-team/issues/13185)

abhishekbhatia1710 committed:
[Security Solution][Entity Analytics][PrivMon] Add label support to JSON API for privileged user monitoring

- Add entity_analytics_monitoring.labels field to API schemas
- Implement source-aware label management in CRUD service
- Support creating/updating users with API labels
- Implement intelligent label merging with existing CSV labels
- Add source-aware deletion logic (only removes API labels when other sources exist)
- Add comprehensive test coverage for all label scenarios

Fixes: EA-13185

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint_all_files --no-cache --fix'

abhishekbhatia1710 committed:
Fix OpenAPI schema validation errors

- Fix YAML indentation issues in common.schema.yaml
- Properly structure entity_analytics_monitoring.labels field
- Ensure all schemas are properly nested under components.schemas
- Regenerate OpenAPI types successfully

Fixes CI failures: Check OAS Snapshot and Check Types

abhishekbhatia1710 committed:
Merge branch 'ea-13185-privmon-labels-json-api' of github.com:abhishekbhatia1710/kibana into ea-13185-privmon-labels-json-api

abhishekbhatia1710 committed:
Merge branch 'main' into ea-13185-privmon-labels-json-api

kibanamachine committed:
[CI] Auto-commit changed files from 'yarn openapi:bundle'

kibanamachine committed:
[CI] Auto-commit changed files from 'make api-docs'

natasha-moore-elastic reviewed (approved):
Approving from the API docs side to unblock the PR, but ideally the newly added properties should have descriptions.
Thanks!

abhishekbhatia1710 committed:
- Adding description to the newly added properties
- Adding _index to the test

abhishekbhatia1710 committed:
Merge branch 'ea-13185-privmon-labels-json-api' of github.com:abhishekbhatia1710/kibana into ea-13185-privmon-labels-json-api

CAWilson94 commented on x-pack/solutions/security/plugins/security_solution/server/lib/entity_analytics/privilege_monitoring/users/privileged_users_crud.ts:
```diff
@@ -47,6 +47,28 @@ export const createPrivilegedUsersCrudService = ({
+              mergedLabels[existingLabelIndex] = { ...apiLabel, source: 'api' };
+            } else {
+              // Add new API label
+              mergedLabels.push({ ...apiLabel, source: 'api' });
```
Nit: same as delete, could we extract into its own upsert helper?

abhishekbhatia1710 committed:
Merge branch 'main' into ea-13185-privmon-labels-json-api

abhishekbhatia1710 committed:
Fix @timestamp field positioning in OpenAPI schemas

- Revert unintended @timestamp field movement from top-level to event object
- Preserve entity_analytics_monitoring label support additions
- Add descriptions to newly added label properties
- Fixes positioning issue introduced by CI auto-commit

abhishekbhatia1710 committed:
Merge branch 'ea-13185-privmon-labels-json-api' of github.com:abhishekbhatia1710/kibana into ea-13185-privmon-labels-json-api

CAWilson94 commented on x-pack/solutions/security/plugins/security_solution/server/lib/entity_analytics/privilege_monitoring/users/privileged_users_crud.ts:
```diff
@@ -47,6 +47,28 @@ export const createPrivilegedUsersCrudService = ({
             ? existingSources
             : [...existingSources, source];
 
+          // Handle API labels if provided
```
For new readers, it’s a bit tricky to quickly see the scenarios being handled.
Essentially it’s: if the user exists, update the labels; otherwise, create a new user—but the method is long enough that it takes a moment to piece that together.

A short top-level comment describing the two branches (or small comments above each section, e.g.
line 30: “Update existing user” and
line 95/96: “Create new user if none exists”)
would make the flow much clearer at a glance.

CAWilson94 reviewed (approved):
Awesome addition! Thanks for this Abhishek 🚀 
I left a couple small comments. 💬

abhishekbhatia1710 committed:
Refactor label merging logic into reusable helper functions

- Extract upsertApiLabel helper function for single label upsert operations
- Extract mergeApiLabels helper function for merging API labels with existing labels
- Remove duplicate label merging code from create and update functions
- Improve code maintainability and follow DRY principle
- Address reviewer feedback for better code organization

Addresses reviewer comments about extracting label upsert logic into reusable helpers.

abhishekbhatia1710 committed:
Add clear documentation to create method for better readability

- Add top-level comment explaining the two main scenarios handled
- Add section comments: 'Update existing user' and 'Create new user if none exists'
- Improve code flow clarity for new readers
- Make it easier to understand the method's logic at a glance

Addresses reviewer feedback about method clarity and documentation.

kibanamachine committed:
[CI] Auto-commit changed files from 'yarn openapi:bundle'

kibanamachine committed:
[CI] Auto-commit changed files from 'make api-docs'

kibanamachine committed:
[CI] Auto-commit changed files from 'yarn openapi:generate'

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint_all_files --no-cache --fix'

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/340442)
* Commit: d48a2dec89a6c6ba480b7526f230b45949ba2240

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/340442#0199582c-f118-43e5-8b44-6fef7b10e498)

### History
* :broken_heart: [Build #339874](https://buildkite.com/elastic/kibana-pull-request/builds/339874) failed cd2e50eba71e9ac759410a73b55cbe3ddbf3a288
* :broken_heart: [Build #339785](https://buildkite.com/elastic/kibana-pull-request/builds/339785) failed e92e6f01277b8f58ee28d0759845624ea985b74a
* :broken_heart: [Build #339764](https://buildkite.com/elastic/kibana-pull-request/builds/339764) failed ea902f680a1c308e1974aefcc3d3d62f69a52f88

cc @abhishekbhatia1710

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340442","number":340442,"commit":"d48a2dec89a6c6ba480b7526f230b45949ba2240"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339874","number":339874,"commit":"cd2e50eba71e9ac759410a73b55cbe3ddbf3a288"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339785","number":339785,"commit":"e92e6f01277b8f58ee28d0759845624ea985b74a"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339764","number":339764,"commit":"ea902f680a1c308e1974aefcc3d3d62f69a52f88"}],"number":340442}
buildkite-pr-comment-->

abhishekbhatia1710 committed:
Fixing types

abhishekbhatia1710 committed:
Merge branch 'ea-13185-privmon-labels-json-api' of github.com:abhishekbhatia1710/kibana into ea-13185-privmon-labels-json-api

abhishekbhatia1710 committed:
Merge branch 'main' into ea-13185-privmon-labels-json-api

abhishekbhatia1710 committed:
Merge branch 'main' into ea-13185-privmon-labels-json-api

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/340725)
* Commit: 62c8d99455388a1d6861d3efe7c89e733412b36f

### Failed CI Steps
* [Jest Tests #9](https://buildkite.com/elastic/kibana-pull-request/builds/340725#01995c58-7fb1-4af0-961d-dff1e2188b54)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/340725#01995c58-7fb1-4af0-961d-dff1e2188b54) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/340725/jobs/01995c58-7fb1-4af0-961d-dff1e2188b54/artifacts/01995c7c-9900-4572-a942-ebf5554bf7c1) Jest Tests #<span></span>9 / createPrivilegedUsersCrudService create should update existing user when username already exists
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/340725#01995c58-7fb1-4af0-961d-dff1e2188b54) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/340725/jobs/01995c58-7fb1-4af0-961d-dff1e2188b54/artifacts/01995c7c-98fd-42db-a4ce-84310e7d439c) Jest Tests #<span></span>9 / createPrivilegedUsersCrudService delete should delete user by id
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/340725#01995c58-7fb1-4af0-961d-dff1e2188b54) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/340725/jobs/01995c58-7fb1-4af0-961d-dff1e2188b54/artifacts/01995c7c-9900-4384-a807-2ec21a48d791) Jest Tests #<span></span>9 / createPrivilegedUsersCrudService error handling should propagate elasticsearch errors on delete
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/340725#01995c58-7fb1-4af0-961d-dff1e2188b54) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/340725/jobs/01995c58-7fb1-4af0-961d-dff1e2188b54/artifacts/01995c7c-98ff-4887-b8ee-bb5d7a17ac4a) Jest Tests #<span></span>9 / createPrivilegedUsersCrudService error handling should propagate elasticsearch errors on update
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/340725#01995c58-7fb1-4af0-961d-dff1e2188b54) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/340725/jobs/01995c58-7fb1-4af0-961d-dff1e2188b54/artifacts/01995c7c-98ff-46f0-a636-f0057d469915) Jest Tests #<span></span>9 / createPrivilegedUsersCrudService label functionality update with labels should update user with new API labels
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/340725#01995c58-7fb1-4af0-961d-dff1e2188b54) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/340725/jobs/01995c58-7fb1-4af0-961d-dff1e2188b54/artifacts/01995c7c-9900-478c-91a3-59a45a3008e3) Jest Tests #<span></span>9 / createPrivilegedUsersCrudService update should update user and return updated document

### History
* :broken_heart: [Build #340681](https://buildkite.com/elastic/kibana-pull-request/builds/340681) failed d81e6c5c79aa2c92048878247edea64184b71789
* :broken_heart: [Build #340442](https://buildkite.com/elastic/kibana-pull-request/builds/340442) failed d48a2dec89a6c6ba480b7526f230b45949ba2240
* :broken_heart: [Build #339874](https://buildkite.com/elastic/kibana-pull-request/builds/339874) failed cd2e50eba71e9ac759410a73b55cbe3ddbf3a288
* :broken_heart: [Build #339785](https://buildkite.com/elastic/kibana-pull-request/builds/339785) failed e92e6f01277b8f58ee28d0759845624ea985b74a
* :broken_heart: [Build #339764](https://buildkite.com/elastic/kibana-pull-request/builds/339764) failed ea902f680a1c308e1974aefcc3d3d62f69a52f88

cc @abhishekbhatia1710

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340725","number":340725,"commit":"62c8d99455388a1d6861d3efe7c89e733412b36f"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340681","number":340681,"commit":"d81e6c5c79aa2c92048878247edea64184b71789"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340442","number":340442,"commit":"d48a2dec89a6c6ba480b7526f230b45949ba2240"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339874","number":339874,"commit":"cd2e50eba71e9ac759410a73b55cbe3ddbf3a288"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339785","number":339785,"commit":"e92e6f01277b8f58ee28d0759845624ea985b74a"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339764","number":339764,"commit":"ea902f680a1c308e1974aefcc3d3d62f69a52f88"}],"number":340725}
buildkite-pr-comment-->

abhishekbhatia1710 committed:
Fixed the test case to include the new entity_analytics_monitoring field in the update and delete operations.

abhishekbhatia1710 committed:
Fixing types

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/341232)
* Commit: 785e6f23c2abb2484898e5339859614cf8c2d4d0

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/341232#019960f6-ade0-47d3-ad59-6eb68b96e67e)

### History
* :broken_heart: [Build #340861](https://buildkite.com/elastic/kibana-pull-request/builds/340861) failed c10ff63f938374a59512bfcea2990e4ce229741a
* :broken_heart: [Build #340725](https://buildkite.com/elastic/kibana-pull-request/builds/340725) failed 62c8d99455388a1d6861d3efe7c89e733412b36f
* :broken_heart: [Build #340681](https://buildkite.com/elastic/kibana-pull-request/builds/340681) failed d81e6c5c79aa2c92048878247edea64184b71789
* :broken_heart: [Build #340442](https://buildkite.com/elastic/kibana-pull-request/builds/340442) failed d48a2dec89a6c6ba480b7526f230b45949ba2240
* :broken_heart: [Build #339874](https://buildkite.com/elastic/kibana-pull-request/builds/339874) failed cd2e50eba71e9ac759410a73b55cbe3ddbf3a288

cc @abhishekbhatia1710

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/341232","number":341232,"commit":"785e6f23c2abb2484898e5339859614cf8c2d4d0"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340861","number":340861,"commit":"c10ff63f938374a59512bfcea2990e4ce229741a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340725","number":340725,"commit":"62c8d99455388a1d6861d3efe7c89e733412b36f"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340681","number":340681,"commit":"d81e6c5c79aa2c92048878247edea64184b71789"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340442","number":340442,"commit":"d48a2dec89a6c6ba480b7526f230b45949ba2240"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/339874","number":339874,"commit":"cd2e50eba71e9ac759410a73b55cbe3ddbf3a288"}],"number":341232}
buildkite-pr-comment-->

abhishekbhatia1710 committed:
Fixing types

abhishekbhatia1710 committed:
Merge branch 'main' into ea-13185-privmon-labels-json-api

abhishekbhatia1710 committed:
Merge branch 'ea-13185-privmon-labels-json-api' of github.com:abhishekbhatia1710/kibana into ea-13185-privmon-labels-json-api

abhishekbhatia1710 committed:
Resolve merge conflicts in privilege monitoring files

kibanamachine committed:
Changes from yarn openapi:bundle

kibanamachine committed:
Changes from make api-docs

abhishekbhatia1710 committed:
Fixing tests and types

abhishekbhatia1710 committed:
Merge branch 'ea-13185-privmon-labels-json-api' of github.com:abhishekbhatia1710/kibana into ea-13185-privmon-labels-json-api

abhishekbhatia1710 committed:
Merge remote-tracking branch 'upstream/main' into ea-13185-privmon-labels-json-api

abhishekbhatia1710 commented:
> @abhishekbhatia1710 is this PR ready to merge in?

Thanks! Yes, it should be. I’ve merged main and turned on auto-merge, so it will merge automatically once CI passes.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/365490)
* Commit: 2b5444ea6f94c43d9fd0ecb1c13965ac25ebee73

### Failed CI Steps
* [Defend Workflows Cypress Tests on Serverless #2](https://buildkite.com/elastic/kibana-pull-request/builds/365490#019abefd-bea8-42ba-b99a-7c6b20b365b2)

### History
* :broken_heart: [Build #356622](https://buildkite.com/elastic/kibana-pull-request/builds/356622) failed 490b8dca79a03e100407f2c45fc4329b691dede1
* :broken_heart: [Build #345030](https://buildkite.com/elastic/kibana-pull-request/builds/345030) failed 2f30b791cc8ed9753b4703262ea2f88ba3684ae0
* :broken_heart: [Build #341232](https://buildkite.com/elastic/kibana-pull-request/builds/341232) failed 785e6f23c2abb2484898e5339859614cf8c2d4d0
* :broken_heart: [Build #340861](https://buildkite.com/elastic/kibana-pull-request/builds/340861) failed c10ff63f938374a59512bfcea2990e4ce229741a
* :broken_heart: [Build #340725](https://buildkite.com/elastic/kibana-pull-request/builds/340725) failed 62c8d99455388a1d6861d3efe7c89e733412b36f

cc @abhishekbhatia1710

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/365490","number":365490,"commit":"2b5444ea6f94c43d9fd0ecb1c13965ac25ebee73"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/356622","number":356622,"commit":"490b8dca79a03e100407f2c45fc4329b691dede1"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/345030","number":345030,"commit":"2f30b791cc8ed9753b4703262ea2f88ba3684ae0"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/341232","number":341232,"commit":"785e6f23c2abb2484898e5339859614cf8c2d4d0"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340861","number":340861,"commit":"c10ff63f938374a59512bfcea2990e4ce229741a"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/340725","number":340725,"commit":"62c8d99455388a1d6861d3efe7c89e733412b36f"}],"number":365490}
buildkite-pr-comment-->

abhishekbhatia1710 committed:
Merge branch 'main' into ea-13185-privmon-labels-json-api

abhishekbhatia1710 committed:
Merge upstream/main into ea-13185-privmon-labels-json-api

Resolved conflicts in privileged_users_crud.ts:
- Updated function signature to use direct parameters (esClient, index, logger) instead of deps object
- Preserved label merging logic and timestamp handling from feature branch
- Maintained backward compatibility with existing call sites

abhishekbhatia1710 committed:
Merge branch 'ea-13185-privmon-labels-json-api' of github.com:abhishekbhatia1710/kibana into ea-13185-privmon-labels-json-api

abhishekbhatia1710 committed:
Merge branch 'main' into ea-13185-privmon-labels-json-api
